genie-web/src/main/java/com/netflix/genie/web/rpc/grpc/services/impl/v4/GRpcJobFileSyncServiceImpl.java

```diff
@@ -0,0 +1,609 @@
+
+                final JobDirectoryState agentJobDirectoryState = syncComplete.getFinalAgentDirectoryState();
+                final boolean includeMd5 = agentJobDirectoryState.getIncludesMd5();
+                final Set<JobFileState> agentJobFileStates = agentJobDirectoryState
+                    .getFilesList()
+                    .stream()
+                    .map(
+                        jobFileState ->
+                            new JobFileState(
+                                jobFileState.getPath(),
+                                jobFileState.getSize(),
+                                includeMd5 ? jobFileState.getMd5() : null
+                            )
+                    )
+                    .collect(Collectors.toSet());
+
+                final Set<JobFileState> serverJobFileStates
+                    = this.jobFileService.getJobDirectoryFileState(this.jobId, includeMd5);
+
+                if (!agentJobFileStates.equals(serverJobFileStates)) {
+                    log.warn(
+                        "After the agent finished syncing job files for job {} the state of the files on the "
+                            + "server {} is different than the supplied state of the files on the agent {}",
+                        this.jobId,
+                        serverJobFileStates,
+                        agentJobFileStates
+                    );
+
+                    // TODO: Probably invoke some service API which will in the background download from long term
+                    //       storage location into a tmp location and then swap it into place for existing
+                    //       job directory locally
+                }
+
+                // TODO: Should invoke some log upload service here to backup the completed logs somewhere OR
+                //       if we've been pushing logs to long term archival in the background signal that these are done
+                //       and can be closed
+            }
+        }
+
+        private void sendResetMessageIfNecessary() {
+            if (!this.sentResetMessage) {
+                log.debug("Sending job file sync reset message to agent");
+                this.responseObserver.onNext(
+                    SyncResponse
+                        .newBuilder()
+                        .setReset(ResetSync.newBuilder().build())
+                        .build()
+                );
+                this.sentResetMessage = true;
+            }
+        }
+
+        private void checkIfShouldSendAck() {
+            boolean sendAck = false;
+            synchronized (this.messagesLock) {
+                if (this.successfulSyncIds.size() + this.failedSyncIds.size() == this.maxSyncMessages) {
+                    sendAck = true;
+                }
+            }
+
+            if (sendAck) {
+                // Proactively send ACK don't wait for timed thread
+                this.sendSyncAckMessageIfNecessary();
+            }
+        }
+    }
+}
```

Comment:
This protocol is stateful on the server, if it gets any more complex we may consider using spring statemachine to encode it.

Reply:
I agree we may want to do that or perhaps should. Outside the scope of this PR but it sounds like it might be the right thing to do.